File: golang/isPalindromeString.go
package main

import (
	"regexp"
	"strings"
)

func IsPalindromeString(text string) bool {
	re := regexp.MustCompile(`[^a-zA-Z0-9]+`)
	cleaned := strings.ToLower(re.ReplaceAllString(text, ""))
	runes := []rune(cleaned)

	for i, j := 0, len(cleaned)-1; i < j; i, j = i+1, j-1 {
		runes[i], runes[j] = runes[j], runes[i]
	}

	return cleaned == string(runes)
}

File: typescript/isPalindromeString.ts
function isPalindromeString(text: string) {
  const string1 = text.replace(/[^a-zA-Z0-9]/g, "").toLowerCase()
  const string2 = text.replace(/[^a-zA-Z0-9]/g, "").toLowerCase().split("").reverse().join("")
  return string1 == string2
}

function isPalindromeString2(text: string) {
  const cleaned = text.replace(/[^a-zA-Z0-9]/g, "").toLowerCase()
  const chars: string[] = Array.from(cleaned)
  for (let i = 0, j = cleaned.length - 1; i < j; i++, j--) {
    [chars[i], chars[j]] = [chars[j]!, chars[i]!]
  }
  return cleaned == chars.join("")
}

console.log(isPalindromeString("A man, a plan, a canal - Panama!"))
console.log(isPalindromeString2("A man, a plan, a canal - Panama!"))
